### PR TITLE
Begin refactoring configuration module

### PR DESCRIPTION
**Pull request by jtgeibel**
This commit series begins refactoring the loading of configuration settings from environment variables. Unlike the main server binary, most binaries just need access to the database, and sometimes S3.

There are still some modules that access the environment directly (such as `app`, `emails`, `git`, some middleware, etc.). This should lay the groundwork for future consolidation of this logic under the `config` module.

r? @Turbo87

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -30,6 +30,12 @@ pub struct Server {
     pub instance_metrics_log_every_seconds: Option<u64>,
 }
 
+#[derive(Debug)]
+pub struct DbPoolConfig {
+    pub url: String,
+    pub read_only_mode: bool,
+}
+
```
Comment (Turbo87) [inline src/config.rs:37]: I assume this part was unintentionally added again in this commit?

File: src/config.rs [hunk 2]
```diff
@@ -41,21 +47,19 @@ impl Default for Server {
...
-    /// - `S3_BUCKET`: The S3 bucket used to store crate files. If not present during development,
-    ///    cargo_registry will fall back to a local uploader.
-    /// - `S3_REGION`: The region in which the bucket was created. Optional if US standard.
-    /// - `S3_ACCESS_KEY`: The access key to interact with S3. Optional if running a mirror.
-    /// - `S3_SECRET_KEY`: The secret key to interact with S3. Optional if running a mirror.
     /// - `SESSION_KEY`: The key used to sign and encrypt session cookies.
     /// - `GH_CLIENT_ID`: The client ID of the associated GitHub application.
     /// - `GH_CLIENT_SECRET`: The client secret of the associated GitHub application.
+    /// - `DATABASE_URL`: The URL of the postgres database to use.
+    /// - `READ_ONLY_REPLICA_URL`: The URL of an optional postgres read-only replica database.
```
Comment (Turbo87) [inline src/config.rs]: same question for these

**Review by Turbo87 — COMMENTED** [review]
LGTM, just one small issue in the second commit, which looks like it unintentionally reverts some of the changes of the first commit

**Comment by jtgeibel** [comment]
Great catch! It looks like I accidentally reintroduced those lines during a rebase. I've resolved that and fixed the formatting issue as well.

File: src/config.rs [hunk 3]
```diff
@@ -67,86 +60,10 @@ impl Default for Server {
     ///
     /// # Panics
     ///
-    /// This function panics if the Server configuration is invalid.
+    /// This function panics if `DB_OFFLINE=leader` but `READ_ONLY_REPLICA_URL` is unset.
```
Comment (Turbo87) [inline src/config.rs:61]: this also looks like an unintentional change, right?

Comment (jtgeibel) [reply src/config.rs]: Yup, fixed

**Comment by Turbo87** [comment]
👏 

@bors r+

**Pull request merged**